MatrixInterpolator.C

```diff
@@ -0,0 +1,408 @@
+                }
+
+                // Invert matrix.
+                log_ev->inverse(log_ev_inv);
+
+                // Perform log mapping.
+                Matrix* log_ev_mult_log_eig = log_ev->mult(log_eigs);
+                Matrix* gamma = log_ev_mult_log_eig->mult(log_ev_inv);
+                gammas.push_back(gamma);
+
+                delete log_ev;
+                delete log_ev_inv;
+                delete log_eigs;
+                delete log_ev_mult_log_eig;
+            }
+        }
+
+        delete x_half_power_inv;
+
+        // Obtain lambda for the P interpolation matrix
+        obtainLambda(gammas);
+    }
+
+    // Obtain distances from database points to new point
+    std::vector<double> inv_q = obtainRBF(point);
+
+    // Interpolate gammas to get gamma for new point
+    Matrix* log_interpolated_matrix = obtainLogInterpolatedMatrix(inv_q);
+
+    // Diagonalize the new gamma so we can exponentiate it
+    EigenPair exp_eigenpair = SymmetricRightEigenSolve(log_interpolated_matrix);
+    delete log_interpolated_matrix;
+
+    Matrix* exp_ev = exp_eigenpair.ev;
+    Matrix* exp_ev_inv = NULL;
+
+    Matrix* exp_eigs = new Matrix(exp_eigenpair.eigs.size(), exp_eigenpair.eigs.size(), false);
+    for (int i = 0; i < exp_eigenpair.eigs.size(); i++)
+    {
+        exp_eigs->item(i, i) = std::exp(exp_eigenpair.eigs[i]);
+    }
+
+    exp_ev->inverse(exp_ev_inv);
+    Matrix* exp_ev_mult_exp_eig = exp_ev->mult(exp_eigs);
+
+    // Expoentiate gamma
+    Matrix* exp_gamma = exp_ev_mult_exp_eig->mult(exp_ev_inv);
+
+    delete exp_ev;
+    delete exp_ev_inv;
+    delete exp_eigs;
+    delete exp_ev_mult_exp_eig;
+
+    // Obtain exp mapping by doing X^1/2*exp(gamma)*X^1/2
+    Matrix* x_half_power_mult_exp_gamma = d_x_half_power->mult(exp_gamma);
+    Matrix* interpolated_matrix = x_half_power_mult_exp_gamma->mult(d_x_half_power);
+
+    delete  x_half_power_mult_exp_gamma;
+    return interpolated_matrix;
+
+}
+
+Matrix* MatrixInterpolator::interpolateNonSingularMatrix(Vector* point)
+{
+    if (d_lambda_T == NULL)
+    {
+        // Invert X
+        Matrix* ref_matrix_inv = NULL;
+        d_rotated_reduced_matrices[d_ref_point]->inverse(ref_matrix_inv);
+
+        // Perform log mapping.
+        std::vector<Matrix*> gammas;
+
+        for (int i = 0; i < d_parameter_points.size(); i++)
+        {
+            // For ref_point, gamma is the zero matrix
+            if (i == d_ref_point)
+            {
+                Matrix* gamma = new Matrix(ref_matrix_inv->numRows(), ref_matrix_inv->numColumns(), ref_matrix_inv->distributed());
+                gammas.push_back(gamma);
+            }
+            else
+            {
+                Matrix* y_mult_ref_matrix_inv = d_rotated_reduced_matrices[i]->mult(ref_matrix_inv);
+
+                // Diagonalize YX^-1 to obtain log of this matrix.
+                EigenPair log_eigenpair = SymmetricRightEigenSolve(y_mult_ref_matrix_inv);
```

Comment:
Could you please document how `log` of a `Matrix` is defined (theoretically) and computed in practice? It seems `log` and `exp` could be member functions of `Matrix` (non-distributed case), rather than just blocks of code here in the interpolator.

Reply:
log, exp, and i think square root and many others are calculated the same way. to get the log of a matrix M, diagonalize M, then do

eigenvectors * log/exp/sqrt/etc.(eigenvalues) * inverse(eigenvectors).

so yes, that part can be moved into Matrix. I'll see about how to do that.

Reply:
I'm not sure what conditions ensure that the eigenvalues of `YX^{-1}` are positive, so that log can be taken. There seems to be an assumption that two matrices `X` and `Y` are close enough in some sense that `YX^{-1}` is close to identity, with positive eigenvalues. This seems likely to fail in some cases, so an error message would be helpful, rather than just the `CAROM_VERIFY` below.